src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -316,26 +287,61 @@ public boolean replace(final TransportRequest request, boolean retainMode, Strin
         }, false);
     }
 
+    static final class IndexResolveKey {
+        final IndicesOptions opts;
+        final boolean isSearchOrFieldCapabilities;
+        final String[] original;
+        public IndexResolveKey(IndicesOptions opts, boolean isSearchOrFieldCapabilities, String[] original) {
+            this.opts = opts;
+            this.isSearchOrFieldCapabilities = isSearchOrFieldCapabilities;
+            this.original = original;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            IndexResolveKey that = (IndexResolveKey) o;
+            return isSearchOrFieldCapabilities == that.isSearchOrFieldCapabilities &&
+                    Objects.equal(opts, that.opts) &&
+                    Arrays.equals(original, that.original);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(opts, isSearchOrFieldCapabilities) + 31*Arrays.hashCode(original);
+        }
+    }
+
     public Resolved resolveRequest(final Object request) {
         if(log.isDebugEnabled()) {
             log.debug("Resolve aliases, indices and types from {}", request.getClass().getSimpleName());
         }
 
         final Resolved.Builder resolvedBuilder = new Resolved.Builder();
         final AtomicBoolean isIndicesRequest = new AtomicBoolean();
+
         getOrReplaceAllIndices(request, new IndicesProvider() {
+            // resolve cache helps us big time on bulk requests
+            final HashMap<IndexResolveKey, Resolved> cache = new HashMap<>();
 
             @Override
             public String[] provide(String[] original, Object localRequest, boolean supportsReplace) {
                 final IndicesOptions indicesOptions = indicesOptionsFrom(localRequest);
-                final Resolved iResolved = resolveIndexPatterns(indicesOptions, localRequest, original);
-                resolvedBuilder.add(iResolved);
-                isIndicesRequest.set(true);
-
-                if(log.isTraceEnabled()) {
-                    log.trace("Resolved patterns {} for {} ({}) to {}", original, localRequest.getClass().getSimpleName(), request.getClass().getSimpleName(), iResolved);
+                final boolean isSearchOrFieldCapabilities = localRequest instanceof FieldCapabilitiesRequest || localRequest instanceof SearchRequest;
+                final IndexResolveKey lookup = new IndexResolveKey(indicesOptions, isSearchOrFieldCapabilities, original);
+                // skip the whole thing if we have seen this exact resolveIndexPatterns result
+                if (cache.get(lookup) == null) {
```

Comment:
There is no need to call `cache.get(...)` (double get).

Reply:
In fact it needs to call get, but there is no need to use the result down the line...
Switched to Set and renamed "cache" to resolvedBefore.

Reply:
With `Map` it is not necessary to call both `get()` and `computeIfAbsent()` and I agree that it is not necessary to maintain `Map` and `Set` provides the required functionality.